src/writer/stripe.rs

```diff
@@ -0,0 +1,175 @@
+use std::io::{Seek, Write};
+
+use arrow::array::RecordBatch;
+use arrow::datatypes::{DataType as ArrowDataType, FieldRef, SchemaRef};
+use prost::Message;
+use snafu::ResultExt;
+
+use crate::error::{IoSnafu, Result};
+use crate::proto;
+
+use super::column::{
+    ByteStripeEncoder, ColumnStripeEncoder, DoubleStripeEncoder, EstimateMemory,
+    FloatStripeEncoder, Int16StripeEncoder, Int32StripeEncoder, Int64StripeEncoder,
+};
+use super::{ColumnEncoding, StreamType};
+
+#[derive(Copy, Clone, Eq, Debug, PartialEq)]
+pub struct StripeInformation {
+    pub start_offset: u64,
+    pub index_length: u64,
+    pub data_length: u64,
+    pub footer_length: u64,
+    pub row_count: usize,
+}
+
+impl From<&StripeInformation> for proto::StripeInformation {
+    fn from(value: &StripeInformation) -> Self {
+        proto::StripeInformation {
+            offset: Some(value.start_offset),
+            index_length: Some(value.index_length),
+            data_length: Some(value.data_length),
+            footer_length: Some(value.footer_length),
+            number_of_rows: Some(value.row_count as u64),
+            encrypt_stripe_id: None,
+            encrypted_local_keys: vec![],
+        }
+    }
+}
+
+/// Encode a stripe. Will encode columns into an in-memory buffer before flushing
+/// entire stripe to the underlying writer.
+pub struct StripeWriter<W> {
+    writer: W,
+    /// Flattened columns, in order of their column ID.
+    columns: Vec<Box<dyn ColumnStripeEncoder>>,
+    pub row_count: usize,
+}
+
+impl<W> EstimateMemory for StripeWriter<W> {
+    /// Used to estimate when stripe size is over threshold and should be flushed
+    /// to the writer and a new stripe started.
+    fn estimate_memory_size(&self) -> usize {
+        self.columns.iter().map(|c| c.estimate_memory_size()).sum()
+    }
+}
+
+impl<W: Write + Seek> StripeWriter<W> {
+    pub fn new(writer: W, schema: &SchemaRef) -> Self {
+        let columns = schema.fields().iter().map(create_encoder).collect();
+        Self {
+            writer,
+            columns,
+            row_count: 0,
+        }
+    }
+
+    /// Attempt to encode entire [`RecordBatch`]. Relies on caller slicing the batch
+    /// to required batch size.
+    pub fn encode_batch(&mut self, batch: &RecordBatch) -> Result<()> {
+        // TODO: consider how to handle nested types (including parent nullability)
+        for (array, encoder) in batch.columns().iter().zip(self.columns.iter_mut()) {
+            encoder.encode_array(array)?;
+        }
+        self.row_count += batch.num_rows();
+        Ok(())
+    }
+
+    /// Flush streams to the writer, and write the stripe footer to finish
+    /// the stripe. After this, the [`StripeWriter`] will be reset and ready
+    /// to write a fresh new stripe.
+    pub fn finish_stripe(&mut self) -> Result<StripeInformation> {
+        let start_offset = self.writer.stream_position().context(IoSnafu)?;
```

Comment:
If we can maintain the offset ourselves, we can avoid seeking. This is extremely important for storage services like S3.

Reply:
Need for Seek removed from writer :+1: